lib/ui/text/paragraph_builder.cc

```diff
@@ -417,14 +417,25 @@ void ParagraphBuilder::pop() {
   }
 }
 
-void ParagraphBuilder::addText(const std::string& text) {
+void ParagraphBuilder::addText(const std::u16string& text) {
+  if (text.empty())
+    return;
+
+  // Use ICU to validate the UTF-16 input.
+  const UChar* text_ptr = reinterpret_cast<const UChar*>(text.data());
+  UErrorCode error_code;
+  u_strToUTF8(nullptr, 0, nullptr, text_ptr, text.size(), &error_code);
+  if (error_code != U_BUFFER_OVERFLOW_ERROR)
+    Dart_ThrowException(tonic::ToDart("string is not well-formed UTF-16"));
```

Comment:
This does no return and does not unwind the stack. We should avoid using this here and return an error code instead.

Reply:
This is safe here because we just entered from Dart and have not yet done any allocation, but given the complexity we probably should reduce use of Dart_ThrowException in native code.  Will move this up to the Dart side.

Reply:
> but given the complexity we probably should reduce use of Dart_ThrowException in native code.

Agreed. I was looking at other instances where we use `Dart_ThrowException` and there are definitely some cases where objects have been allocated on the stack. We can go over those at a later point.